Index: linux-3.3.8.ralink/drivers/net/wireless/Makefile
===================================================================
--- linux-3.3.8.ralink/drivers/net/wireless/Makefile	(版本 30)
+++ linux-3.3.8.ralink/drivers/net/wireless/Makefile	(工作副本)
@@ -61,3 +61,35 @@
 
 obj-$(CONFIG_BRCMFMAC)	+= brcm80211/
 obj-$(CONFIG_BRCMSMAC)	+= brcm80211/
+
+ifneq ($(CONFIG_RT2860V2_AP),)
+obj-$(CONFIG_RT2860V2_AP)       += rt2860v2_ap/
+endif
+ifneq ($(CONFIG_RT2860V2_STA),)
+obj-$(CONFIG_RT2860V2_STA)      += rt2860v2_sta/
+endif
+ifneq ($(CONFIG_RT3090_AP),)
+obj-$(CONFIG_RT3090_AP)      	+= RT3090_ap/
+endif
+ifneq ($(CONFIG_RT5392_AP),)
+obj-$(CONFIG_RT5392_AP)      	+= RT5392_ap/
+endif
+ifneq ($(CONFIG_RT3572_AP),)
+obj-$(CONFIG_RT3572_AP)      	+= RT3572_ap/
+endif
+ifneq ($(CONFIG_RT5572_AP),)
+obj-$(CONFIG_RT5572_AP)      	+= RT5572_ap/
+endif
+ifneq ($(CONFIG_RT305x_INIC_MII),)
+obj-$(CONFIG_RT2880v2_INIC_MII) += iNIC/mii/
+endif
+ifneq ($(CONFIG_RT305x_INIC_MII),)
+obj-$(CONFIG_RT2880v2_INIC_PCI) += iNIC/pci/
+endif
+
+ifneq ($(CONFIG_RT305x_INIC_MII),)
+obj-$(CONFIG_RT305x_INIC_MII) += iNIC_RT305x/mii/
+endif
+ifneq ($(CONFIG_RT305x_INIC_USB),)
+obj-$(CONFIG_RT305x_INIC_USB) += iNIC_RT305x/usb/
+endif
